tests/utils.rs

```diff
@@ -149,3 +208,211 @@ pub fn copy_dir_all(src: impl AsRef<Path>, dst: impl AsRef<Path>) -> Result<()>
     }
     Ok(())
 }
+
+pub fn git_clone(repo_url: &str, target_dir: &Path, revision: Option<&str>) -> Result<()> {
+    if target_dir.exists() {
+        return Err(anyhow!(
+            "Error cloning {}. Target directory '{}' already exists",
+            repo_url,
+            target_dir.display()
+        ));
+    }
+
+    let command = format!("git clone \"{}\" \"{}\"", repo_url, target_dir.display());
+    let command_vec = shell_words::split(command.as_str()).map_err(anyhow::Error::new)?;
+
+    Command::new(&command_vec[0])
+        .args(&command_vec[1..])
+        .stdout(Stdio::null())
+        .status()
+        .with_context(|| format!("Failed to clone repository from {}", repo_url))?;
+
+    if let Some(revision_str) = revision {
+        let result = git_to_revision(target_dir, "origin", revision_str);
+        if let Err(e) = result {
+            // Cleanup! If we cannot checkout the revision, remove the directory.
+            fs::remove_dir_all(target_dir)
+                .with_context(|| format!("Failed to remove directory {}", target_dir.display()))?;
+            return Err(e);
+        }
+    }
+
+    Ok(())
+}
+
+// Resolvees the SHA1 of revision at remote_name.
+// revision can be a tag, a branch, or a commit SHA1.
+fn git_resolve_revision(repo_path: &Path, remote_name: &str, revision: &str) -> Result<String> {
```

Comment:
It seems like now is a good opportunity to carve out an internal package that both tests & Tinty itself can use for shared concerns like these. 

Ideally, a package that lives entirely in this repo. I don't contribute to any other Rust project but I assume such a dependency structure is possible. Let me know if I'm wrong 😅

Reply:
Yeah you're totally right. I'll set it up so we have a shared lib to use between the tests and cli tool.